File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, d):
    answer = 0
    d_num = len(d)
    
    for p in permutations(d,d_num):
        cnt = 0
        hp = k
        for i in p:
            if hp>=i[0]:
                hp-=i[1]
                cnt+=1
        if cnt>answer:
            answer=cnt
        
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    static int answer = 0;
    
    public int solution(int k, int[][] dungeons) {
        boolean[] visited = new boolean[dungeons.length];
        backtrack(k, dungeons, visited, 0);
        
        return answer;
    }
    
    private void backtrack(int k, int[][] dungeons, boolean[] visited, int count) {
        answer = Math.max(answer, count);
        for (int i=0;i<dungeons.length;i++) {
            if (!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;
                backtrack(k-dungeons[i][1], dungeons, visited, count +1);
                visited[i] = false;
            }
        }
        
        
    }
}